### Read the .csv file

In [2]:
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import pandas as pd
df = pd.read_csv('Student Performance.csv')

### KNN Model 5

<img src="Model 5.jpg" width=500 height=500 />

### Obtain the feature matrix and label

In [3]:
# convert these data into 2D numpy array
# order: math percentage, reading score percentage, writing score percentage
X = np.array([[df['math percentage'][0], 
              df['reading score percentage'][0], 
              df['writing score percentage'][0]]], 
             dtype = 'float')
for i in range(1,1000):
    X = np.append(X, [[df['math percentage'][i], 
                  df['reading score percentage'][i], 
                  df['writing score percentage'][i]]],axis=0)
print('feature matrix')
print(X)
print('size: ' + str(X.shape))
print()

# Group A to Group E labeled as 0,1,2,3,4
if(df['race/ethnicity'][0] == 'group A'):
    first_elem = 0
elif(df['race/ethnicity'][0] == 'group B'):
    first_elem = 1
elif(df['race/ethnicity'][0] == 'group C'):
    first_elem = 2
elif(df['race/ethnicity'][0] == 'group D'):
    first_elem = 3
else:
    first_elem = 4
    
y = np.array([first_elem], dtype = 'int')
for i in range(1,1000):
    if(df['race/ethnicity'][i] == 'group A'):
        y = np.append(y, 0)
    elif(df['race/ethnicity'][i] == 'group B'):
        y = np.append(y, 1)
    elif(df['race/ethnicity'][i] == 'group C'):
        y = np.append(y, 2)
    elif(df['race/ethnicity'][i] == 'group D'):
        y = np.append(y, 3)
    else:
        y = np.append(y, 4)
print('Labels')
print(y)
print('size: ' + str(y.shape))


feature matrix
[[0.72 0.72 0.74]
 [0.69 0.9  0.88]
 [0.9  0.95 0.93]
 ...
 [0.59 0.71 0.65]
 [0.68 0.78 0.77]
 [0.77 0.86 0.86]]
size: (1000, 3)

Labels
[1 0 1 1 1 1 0 1 0 1 1 1 1 0 1 1 1 1 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 0 1 1 1 1 0 1 1 0 1 0 0 1 0 1 1 0 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 0 0 1 1
 1 1 1 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0 0 0 1 1 1 0 1 0 0 1 1 0 1 1 0
 1 1 1 0 1 1 1 1 1 0 0 0 1 1 1 1 0 1 1 1 0 0 0 0 1 0 1 1 0 1 1 0 1 1 1 1 1
 0 0 0 1 0 1 1 0 0 1 0 1 0 0 1 1 1 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 1 1 1 0
 0 1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 1 1 1 0 1 0 1 1 0 1 1 0 0 0 0 1 1 0 0 1
 1 1 0 1 1 1 1 0 1 1 1 1 0 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1 1 1 0 1
 0 0 0 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 1 1 1 1 0 1 0 0 1 1 0 1 1 0 1 0 1
 0 0 0 1 0 1 0 1 0 1 0 1 1 1 1 1 0 0 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0
 1 1 1 0 1 1 0 1 1 0 0 1 1 1 0 1 1 1 1 0 1 1 1 1 0 1 1 1 0 1 1 1 0 0 1 1 1
 1 0 1 0 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 1 0 0 1 1 1 1 1 0 1 1 0 1 1 0
 1 0 0 0 0 1 0 0 1 0 1

In [4]:
# Randomly splitting the original dataset into training set and testing set
# The function"train_test_split" from "sklearn.cross_validation" library performs random splitting.
# "test_size=0.3" means that pick 30% of data samples for testing set, and the rest (70%) for training set.

from sklearn.model_selection import train_test_split
# where X is all the sample data, y is all the labels
# random_state=1 means only random once and rest stay the same
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=6) # We can fix the random_state for reproducibility!

In [5]:
# print the size of the traning set:
print(X_train.shape)
print(y_train.shape)

(700, 3)
(700,)


In [6]:
# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)

(300, 3)
(300,)


In [7]:
print(X_test)
print('\n')
print(y_test)

[[0.55 0.41 0.48]
 [0.44 0.54 0.53]
 [0.7  0.64 0.7 ]
 [0.6  0.59 0.54]
 [0.7  0.56 0.51]
 [0.92 0.79 0.84]
 [0.64 0.79 0.77]
 [0.86 0.81 0.75]
 [0.99 0.93 0.9 ]
 [0.78 0.77 0.77]
 [0.73 0.76 0.78]
 [0.33 0.41 0.43]
 [0.69 0.58 0.53]
 [0.65 0.78 0.82]
 [0.45 0.59 0.64]
 [0.59 0.65 0.66]
 [0.97 0.92 0.86]
 [0.57 0.69 0.68]
 [0.81 0.8  0.76]
 [0.63 0.6  0.57]
 [0.54 0.59 0.5 ]
 [0.62 0.67 0.69]
 [0.64 0.66 0.59]
 [0.81 0.72 0.77]
 [0.73 0.76 0.78]
 [0.36 0.29 0.27]
 [0.58 0.61 0.52]
 [0.5  0.48 0.53]
 [0.81 0.88 0.9 ]
 [0.62 0.68 0.68]
 [0.68 0.54 0.53]
 [0.53 0.71 0.67]
 [0.18 0.32 0.28]
 [0.69 0.58 0.57]
 [0.4  0.52 0.43]
 [0.63 0.55 0.63]
 [0.66 0.63 0.64]
 [0.65 0.7  0.71]
 [0.66 0.65 0.6 ]
 [0.61 0.47 0.56]
 [0.43 0.62 0.61]
 [0.5  0.48 0.42]
 [0.82 0.85 0.87]
 [0.6  0.7  0.74]
 [0.87 0.73 0.72]
 [0.82 0.85 0.86]
 [0.88 0.99 0.95]
 [0.64 0.73 0.68]
 [0.69 0.65 0.74]
 [0.66 0.72 0.7 ]
 [0.37 0.45 0.38]
 [0.52 0.57 0.56]
 [0.87 0.84 0.76]
 [0.52 0.7  0.62]
 [0.79 0.73 0.67]
 [0.78 0.8

In [8]:
from sklearn.metrics import accuracy_score
highest = 0
best_k = 0
for i in range(1,300):
    k = i
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
#     print('k=' + str(i) + ', accuracy=' + str(accuracy))
    if(highest == 0):
        highest = accuracy
    elif(highest < accuracy):
        highest = accuracy
        best_k = i
    else:
        pass
print("The highest accuracy: " + str(highest))
print("When k= " + str(best_k))

The highest accuracy: 0.6833333333333333
When k= 64


#### Using Cross Validation on KNN algorithm to see if the model accuary can be improved

In [9]:
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

highest = 0
best_k = 0
for i in range(1,300):
    k = i
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X, y)
    scores = cross_val_score(knn, X, y, cv = 10, scoring = 'accuracy')
    accuracy = scores.mean()
#     y_predict = knn.predict(X_test)
#     accuracy = accuracy_score(y_test, y_predict)
    print('k=' + str(i) + ', accuracy=' + str(accuracy))
    if(highest == 0):
        highest = accuracy
    elif(highest < accuracy):
        highest = accuracy
        best_k = i
    else:
        pass
print("The highest accuracy: " + str(highest))
print("When k= " + str(best_k))

k=1, accuracy=0.599
k=2, accuracy=0.5609999999999999
k=3, accuracy=0.641
k=4, accuracy=0.607
k=5, accuracy=0.656
k=6, accuracy=0.635
k=7, accuracy=0.6489999999999999
k=8, accuracy=0.634
k=9, accuracy=0.671
k=10, accuracy=0.6479999999999999
k=11, accuracy=0.666
k=12, accuracy=0.671
k=13, accuracy=0.681
k=14, accuracy=0.6669999999999999
k=15, accuracy=0.674
k=16, accuracy=0.671
k=17, accuracy=0.675
k=18, accuracy=0.687
k=19, accuracy=0.6829999999999999
k=20, accuracy=0.678
k=21, accuracy=0.688
k=22, accuracy=0.6859999999999999
k=23, accuracy=0.6930000000000001
k=24, accuracy=0.6859999999999998
k=25, accuracy=0.6940000000000001
k=26, accuracy=0.695
k=27, accuracy=0.697
k=28, accuracy=0.6990000000000001
k=29, accuracy=0.708
k=30, accuracy=0.6900000000000002
k=31, accuracy=0.6990000000000001
k=32, accuracy=0.696
k=33, accuracy=0.6950000000000001
k=34, accuracy=0.6990000000000001
k=35, accuracy=0.6990000000000001
k=36, accuracy=0.6979999999999998
k=37, accuracy=0.6969999999999998
k=38, accur

## Conclusion
### KNN model accuracy: 0.71 when k=29 with Cross Validation
### KNN model accuracy: 0.68 when k=64 without Cross Validation